tests/test_helpers.py

```diff
@@ -494,6 +499,515 @@ def test_set_content_disposition_bad_param():
                                            **{'foo\x00bar': 'baz'})
 
 
+def test_dummy_cookie_jar(loop):
+    cookie = helpers.SimpleCookie('foo=bar; Domain=example.com;')
+    dummy_jar = helpers.DummyCookieJar(loop=loop)
+    assert len(dummy_jar) == 0
+    dummy_jar.update_cookies(cookie)
+    assert len(dummy_jar) == 0
+    with pytest.raises(StopIteration):
+        next(iter(dummy_jar))
+    assert dummy_jar.filter_cookies(URL("http://example.com/")) is None
+    dummy_jar.clear()
+
+
+# -------------------------------- DigestAuth --------------------------
+
+@asyncio.coroutine
+def test_digest_auth_MD5(loop, test_client):
+    opaque = binascii.hexlify(os.urandom(16))
+    username = 'username'
+    password = 'password'
+
+    def H(x):
+        if isinstance(x, str):
+            x = x.encode()
+        return hashlib.md5(x).hexdigest()
+
+    @asyncio.coroutine
+    def handler(request):
+        realm = 'test@example.com'
+        if 'Authorization' in request.headers:
+            pattern = re.compile(r'digest ', flags=re.IGNORECASE)
+            auth_data = helpers.parse_key_value_list(
+                pattern.sub('', request.headers['Authorization'], count=1)
+            )
+
+            A1 = '%s:%s:%s' % (username, realm, password)
+            A2 = '%s:%s' % (request.method, auth_data['uri'])
+
+            secret = H(A1)
+            data = '%s:%s:%s:%s:%s' % (
+                auth_data['nonce'],
+                auth_data['nc'],
+                auth_data['cnonce'],
+                auth_data['qop'],
+                H(A2)
+            )
+
+            assert H('%s:%s' % (secret, data)) == auth_data['response']
+
+            return web.Response()
+
+        response = web.Response(status=401)
+        data = '%s:%s' % (time.ctime().encode(), 'secret')
+        nonce = hashlib.sha1(data.encode()).hexdigest()
+        pairs = ', '.join([
+            'realm="%s"' % realm,
+            'qop="auth"',
+            'algorithm=MD5',
+            'nonce="%s"' % nonce,
+            'opaque="%s"' % opaque,
+        ])
+        response.headers['WWW-Authenticate'] = 'Digest %s' % pairs
+        return response
+
+    app = web.Application()
+    app.router.add_route('GET', '/', handler)
+    client = yield from test_client(app)
+
+    auth = helpers.DigestAuth(username, password, client)
+
+    resp = yield from auth.request('GET', '/?test=true')
+    assert 200 == resp.status
+
+
+@asyncio.coroutine
+def test_digest_auth_MD5_sess(loop, test_client):
+    opaque = binascii.hexlify(os.urandom(16))
+    username = 'usr'
```

Comment:
wh usr nd psswd whl username nd password wrwhr ls?

Reply:
I was trying to demonstrate that the helper isn't just hardcoded to work with one username and password. Although, this really doesn't challenge it at all. Once you decide what you want to do about the string escaping I can change this to include something more difficult like spaces and quotes.